x=[20105, 22886, 9914, 48086, 26420, 9024, 49936, 29904, 51833, 18598, 19473, 26450, 19541, 6680, 47879, 29307, 27476, 53240, 10381, 44074, 17139, 48486, 7670, 3058, 19172, 8639, 81774, 26027, 13009, 40227, 142746, 13656, 39003, 14095, 20734, 8401, 29938, 27224, 47698, 23194, 13226, 25767, 1333, 10098, 2572, 3240, 3413, 5433, 2873, 3158, 3865, 10766, 26636, 11715, 14678, 6252, 18539, 24367, 24141, 2788, 18403, 28504, 43727, 25176, 28603, 5039, 13565, 33472, 20645, 4635, 13593, 33726, 31471, 20364, 18348, 5672, 7050, 10179, 27234, 24289, 7811, 12061, 30887, 23400, 21123, 16953, 20621, 12844, 17086, 9895, 19795, 17161, 4472, 6599, 15381, 26428, 8642]
y=[2.584132880796354, 1.8617060413655901, 3.3642164958238054, 1.893821794840749, 2.458618122713424, 1.3823614954475356, 1.5887884988337049, 2.263240452669874, 2.1299312627901674, 3.1167775921366614, 1.843777642559312, 2.1343384717247678, 6.899808440394504, 1.901384588222512, 1.8767971436899227, 2.7518106665006083, 3.2736311330353507, 1.8016921549252645, 3.0739221934632757, 3.0456929514467372, 2.484094537798847, 1.7687761071484012, 2.3395776814452462, 1.9957627118644068, 2.4576398636285464, 1.651110059674472, 2.0314150108214015, 2.0063456645665108, 1.7216014419635108, 2.9356446549779247, 1.9574994122411173, 4.180789597845421, 1.6820887103726458, 3.8776637940903207, 2.343854063423943, 2.54421082614902, 2.9919454908011107, 2.554003978150386, 2.350288731392736, 3.304220214026265, 2.9910822664983177, 2.3575910420961534, 3.3333333333333335, 3.655224973870543, 2.9399346906099564, 2.305922889251836, 3.5293778026905835, 4.58357353106243, 2.4445757600456433, 1.9491268814906004, 5.765883977900552, 1.4537490456194258, 1.71501934020555, 2.385510898798288, 1.5577228422264044, 2.822025837537066, 1.94532459740306, 1.645854373878123, 1.772106220946158, 1.4328000103850513, 4.02711045155153, 1.5303999998846982, 1.3891607147965306, 2.0158657143074215, 3.628910882919799, 2.1668865699194733, 2.6751031048694354, 2.5170904351620695, 1.3136906917033542, 2.6581355606038803, 1.2367246675299512, 2.0222647511547156, 3.4855936922355815, 1.6972898763725632, 1.7173740021613408, 1.2942988361582366, 2.8307592391795433, 1.6979657164475046, 1.6989348127634292, 2.273257152061199, 2.8811954050511948, 1.2261600021796903, 1.7754188260102655, 1.420631221216686, 1.9224592597639232, 3.7449901110691495, 2.0393276258348965, 1.9154735133486134, 1.623025346357531, 3.118153903760807, 1.5741601907544212, 1.3416899063667995, 2.556002598264621, 1.0269195121853822, 1.7169398505412112, 1.5230039987049593, 9.858284217064707]
z=[61, 54, 36, 122, 45, 17, 157, 93, 160, 56, 80, 80, 56, 27, 77, 62, 83, 111, 42, 98, 57, 111, 27, 4, 59, 19, 194, 72, 29, 90, 327, 32, 104, 39, 65, 41, 86, 109, 99, 67, 21, 72, 10, 36, 7, 14, 13, 8, 4, 19, 5, 33, 63, 33, 39, 31, 51, 68, 75, 14, 61, 67, 115, 71, 57, 13, 36, 68, 70, 19, 46, 69, 93, 47, 50, 13, 16, 34, 58, 66, 39, 17, 68, 44, 46, 53, 53, 34, 56, 30, 70, 38, 33, 31, 42, 62, 19]
t=['mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'mavs', 'nlrp3', 'nlrp3', 'nlrp3', 'nlrp3', 'nlrp3', 'nlrp3', 'nlrp3', 'nlrp3', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt', 'wt']

import matplotlib,matplotlib.pyplot
matplotlib.rcParams.update({'font.size':18,'font.family':'Arial','xtick.labelsize':14,'ytick.labelsize':14})
matplotlib.rcParams['pdf.fonttype']=42
import numpy


unique_genotypes=set(t)

for genotype in unique_genotypes:
    a=[]; b=[]; c=[]

    if genotype == 'mavs':
        marker = 'D'
    if genotype == 'nlrp3':
        marker = 's'
    if genotype == 'wt':
        marker = 'o'

    for i in range(len(x)):
        if t[i] == genotype:
            a.append(x[i])
            b.append(y[i])
            c.append(z[i])

    matplotlib.pyplot.scatter(numpy.log10(a),b,c=numpy.log2(c),marker=marker,cmap='viridis',vmin=min(numpy.log2(z)),vmax=max(numpy.log2(z)),label=genotype)

matplotlib.pyplot.axhline(y=1,ls=':',color='black')

cbar=matplotlib.pyplot.colorbar()
cbar.set_ticks([2,3,4,5,6,7,8])
cbar.set_ticklabels(2**(numpy.array([2,3,4,5,6,7,8])))
cbar.ax.set_title('Detected variants',fontsize=14)

matplotlib.pyplot.xlim([3-0.1,numpy.max(numpy.log10(x))+0.1])

matplotlib.pyplot.yticks([1,2,4,6,8,10],['bkg',2,4,6,8,10])

matplotlib.pyplot.xlabel('log$_{10}$ Mapped reads')
matplotlib.pyplot.ylabel('Support ratio')

leg=matplotlib.pyplot.legend(fontsize=14)
for marker in leg.legendHandles:
    marker.set_color('black')

matplotlib.pyplot.tight_layout()
matplotlib.pyplot.savefig('figure.pdf')
matplotlib.pyplot.clf()


